### PR TITLE
MINOR: ConfigKey variable renamed

### PR DESCRIPTION
**Pull request by kamalcph**
- Java doc updated
- Used final modifier where-ever it's applicable
- Unnecessary type-cast removed
- cleanup

### CODE REVIEW

**Comment by cmccabe** [comment]
LGTM (non-binding)

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java [hunk 1]
```diff
@@ -334,14 +334,15 @@ public static void sleep(long ms) {
      * Instantiate the class
      */
     public static <T> T newInstance(Class<T> c) {
+        if (c == null)
+            throw new NullPointerException();
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: Why are we changing the exception type here? That could affect calling code. Also, if there's a good reason to do this, we'd want to include a message in the exception.

File: clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java [hunk 2]
```diff
@@ -158,8 +157,7 @@ public Password getPassword(String key) {
     }
 
     /**
-     * Put all keys that do not start with {@code prefix} and their parsed values in the result map and then
-     * put all the remaining keys with the prefix stripped and their parsed values in the result map.
+     * Put all keys that start with {@code prefix} with the prefix stripped and their parsed values in the result map.
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java]: Why was this changed?

File: clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java [hunk 3]
```diff
@@ -66,52 +66,51 @@ public AbstractConfig(ConfigDef definition, Map<?, ?> originals) {
         this(definition, originals, true);
     }
 
-    protected Object get(String key) {
+    protected <T> T get(String key) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java]: I am not sure about this. It's a really dangerous method as it implicitly casts in an unsafe manner.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR. It improves the code in a few areas. Left a few comments. It would also be good to update the PR title to reflect what's in the PR.

File: clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java (hunk 2 above)
Comment (kamalcph) [reply clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java]: The method doesn't process the keys that don't start with the prefix. So, I have updated the java doc.

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java (hunk 1 above)
Comment (kamalcph) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: I thought it's better to validate the argument before trying to create an instance. And, the code seems unfit in the `try..catch` block.

Comment (kamalcph) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: This won't affect the calling code as both NPE and KafkaException extends RuntimeException. And, the message "Requested class was null" was dumb. 

I did this change, as other methods in this class followed the same pattern.

File: clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java (hunk 3 above)
Comment (kamalcph) [reply clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java]: I agree. Earlier, implicit casts were still done for primitives, list and class ?

File: clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java (hunk 2 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java]: Yes, it does. The copy constructor is used below.

```java
Map<String, Object> result = new RecordingMap<>(values(), prefix, true);
```

File: clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java (hunk 3 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java]: We were doing explicit casts in those cases, right? I agree that it would be better to check before casting if that's what you are suggesting.

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java (hunk 1 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: It makes a difference for code that catches `KafkaException` instead of `RuntimeException`. We'd have to check every usage to make sure it was OK.

Comment (kamalcph) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: Makes sense. I'll revert this change.

File: clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java (hunk 3 above)
Comment (kamalcph) [reply clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java]: Yes, I will do that change in another PR. I'll revert this one too.

File: clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java (hunk 2 above)
Comment (kamalcph) [reply clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java]: Oh, I missed it.

**Comment by kamalcph** [comment]
@ijuma 

Thanks for the detailed review. I've updated this PR. Pl take a look.

**Review by ijuma — APPROVED** [review]
Thanks, LGTM.
